File: resolving/npm-resolver/src/fetchAttestationPublishedAt.ts
import * as retry from '@zkochan/retry'

import type { FetchMetadataFromFromRegistryOptions } from './fetch.js'

/**
 * Per-version publish timestamp from npm's attestation endpoint —
 * `/-/npm/v1/attestations/<name>@<version>`.
 *
 * The response is a small JSON document containing one or more Sigstore
 * bundles. We read `bundle.verificationMaterial.tlogEntries[].integratedTime`
 * (the Rekor inclusion time) and surface it as an ISO date. This is a
 * couple of seconds after the actual publish — close enough for a
 * release-age policy that operates in minutes/hours/days.
 *
 * We deliberately do **not** verify the Sigstore signature here: the
 * trust model is identical to reading the registry's `time` field on
 * the full metadata document. The win is bandwidth — the attestation
 * payload is tens of KB versus the multi-MB full metadata document, so
 * cold-cache + `--frozen-lockfile` installs against a fleet of
 * provenance-published packages pay far less to verify timestamps.
 *
 * Returns `undefined` when:
 *
 * - The package has no published attestations (`404`).
 * - The response is malformed or missing the timestamp.
 * - The request itself fails (network error, registry 5xx).
 *
 * In all of those cases the caller falls back to fetching full metadata.
 */
export interface FetchAttestationOptions {
  registry: string
  authHeaderValue?: string
}

export async function fetchAttestationPublishedAt (
  fetchOpts: FetchMetadataFromFromRegistryOptions,
  pkgName: string,
  version: string,
  opts: FetchAttestationOptions
): Promise<string | undefined> {
  const url = `${opts.registry.replace(/\/$/, '')}/-/npm/v1/attestations/${pkgName}@${version}`
  const retryOperation = retry.operation(fetchOpts.retry)
  return new Promise<string | undefined>((resolve) => {
    retryOperation.attempt(async () => {
      let response: Response
      try {
        response = await fetchOpts.fetch(url, {
          authHeaderValue: opts.authHeaderValue,
          retry: fetchOpts.retry,
          timeout: fetchOpts.timeout,
        })
      } catch {
        // Network errors fall through to the full-metadata path; the
        // caller's `fetchFullMetadataCached` has its own retry policy.
        resolve(undefined)
        return
      }
      // 404 = package never published attestations. Other 4xx/5xx also
      // mean "can't get an answer from this endpoint, fall back."
      if (response.status >= 400) {
        resolve(undefined)
        return
      }
      let body: unknown
      try {
        body = await response.json()
      } catch {
        resolve(undefined)
        return
      }
      resolve(extractPublishedAt(body))
    })
  })
}

/**
 * Pull the earliest `integratedTime` across every attestation bundle in
 * the response and convert it to an ISO timestamp. Earliest is the
 * conservative choice: if two attestations disagree (e.g. publish
 * v0.1 vs SLSA provenance v1), we attribute the publish to the older
 * Rekor entry. The Rekor timestamp is what tells us when the artifact
 * existed in a transparency log — that's the floor on publish time.
 */
function extractPublishedAt (body: unknown): string | undefined {
  if (!body || typeof body !== 'object') return undefined
  const attestations = (body as { attestations?: unknown }).attestations
  if (!Array.isArray(attestations)) return undefined

  let earliestSeconds: number | undefined
  for (const attestation of attestations) {
    const seconds = readEarliestIntegratedTime(attestation)
    if (seconds == null) continue
    if (earliestSeconds == null || seconds < earliestSeconds) {
      earliestSeconds = seconds
    }
  }
  if (earliestSeconds == null) return undefined
  return new Date(earliestSeconds * 1000).toISOString()
}

function readEarliestIntegratedTime (attestation: unknown): number | undefined {
  if (!attestation || typeof attestation !== 'object') return undefined
  const bundle = (attestation as { bundle?: unknown }).bundle
  if (!bundle || typeof bundle !== 'object') return undefined
  const verificationMaterial = (bundle as { verificationMaterial?: unknown }).verificationMaterial
  if (!verificationMaterial || typeof verificationMaterial !== 'object') return undefined
  const tlogEntries = (verificationMaterial as { tlogEntries?: unknown }).tlogEntries
  if (!Array.isArray(tlogEntries)) return undefined

  let earliest: number | undefined
  for (const entry of tlogEntries) {
    if (!entry || typeof entry !== 'object') continue
    const rawIntegratedTime = (entry as { integratedTime?: unknown }).integratedTime
    // npm serializes integratedTime as a string ("1778583836") to avoid
    // JSON precision loss; accept either string or number defensively.
    const seconds = parseIntegratedTimeSeconds(rawIntegratedTime)
    if (seconds == null) continue
    if (earliest == null || seconds < earliest) earliest = seconds
  }
  return earliest
}

function parseIntegratedTimeSeconds (raw: unknown): number | undefined {
  const seconds = typeof raw === 'string' ? Number(raw) : typeof raw === 'number' ? raw : NaN
  if (!Number.isFinite(seconds) || seconds <= 0) return undefined
  return seconds
}

File: pacquet/crates/resolving-npm-resolver/src/fetch_attestation_published_at.rs
//! Per-version publish timestamp from npm's attestation endpoint —
//! `/-/npm/v1/attestations/<name>@<version>`.
//!
//! The endpoint serves a small JSON document containing one or more
//! Sigstore bundles. We read
//! `bundle.verificationMaterial.tlogEntries[].integratedTime`
//! (the Rekor inclusion time) and surface it as an ISO timestamp.
//! This is a few seconds after the actual publish — close enough for
//! a release-age policy that operates in minutes/hours/days, and
//! tens of kilobytes versus the multi-megabyte full-metadata fetch.
//!
//! We deliberately do **not** verify the Sigstore signature here:
//! the trust model is identical to reading the registry's `time`
//! field on the full metadata document.
//!
//! Returns `Ok(None)` (not `Err`) on every "no answer" condition:
//! 4xx/5xx responses, malformed JSON, missing timestamps. The
//! verifier falls back to the next layer of the publish-time lookup
//! chain. Real network errors propagate as
//! [`crate::FetchMetadataError::Network`] so the verifier can fold
//! them into a violation reason instead of swallowing.
//!
//! Verbatim port of upstream's
//! [`fetchAttestationPublishedAt.ts`](https://github.com/pnpm/pnpm/blob/2a9bd897bf/resolving/npm-resolver/src/fetchAttestationPublishedAt.ts).

use chrono::DateTime;
use pacquet_network::{AuthHeaders, ThrottledClient};

use crate::FetchMetadataError;

/// Options bundle for [`fetch_attestation_published_at`].
#[derive(Debug, Clone)]
pub struct FetchAttestationOptions<'a> {
    pub registry: &'a str,
    pub http_client: &'a ThrottledClient,
    pub auth_headers: &'a AuthHeaders,
}

/// Fetch the earliest Rekor `integratedTime` across the attestation
/// bundles for `<name>@<version>`. Returns `Ok(Some(rfc3339))` on a
/// 2xx response with a parseable timestamp; `Ok(None)` for any
/// "no answer" condition (4xx/5xx, malformed body, no timestamps);
/// `Err(_)` only when the underlying request fails before reaching
/// the server.
pub async fn fetch_attestation_published_at(
    pkg_name: &str,
    version: &str,
    opts: &FetchAttestationOptions<'_>,
) -> Result<Option<String>, FetchMetadataError> {
    // Strip a trailing `/` from the registry root before assembling
    // the endpoint URL — `<registry>/-/npm/v1/attestations/...`
    // produces `//` otherwise, which some self-hosted registries
    // reject as a malformed path. Matches upstream's
    // `opts.registry.replace(/\/$/, '')`.
    let registry = opts.registry.trim_end_matches('/');
    let url = format!("{registry}/-/npm/v1/attestations/{pkg_name}@{version}");
    let mut request = opts.http_client.acquire_for_url(&url).await.get(&url);
    if let Some(value) = opts.auth_headers.for_url(&url) {
        request = request.header("authorization", value);
    }
    let response = match request.send().await {
        Ok(response) => response,
        Err(error) => {
            // Mirror upstream's `catch` swallow — return None on
            // network errors so the caller falls through to the
            // full-metadata layer. Surfacing the error would be more
            // informative but inconsistent with upstream.
            tracing::debug!(target: "pacquet_resolving_npm_resolver::attestation", ?error, %url, "attestation fetch failed; falling back");
            return Ok(None);
        }
    };
    // Anything outside the 2xx range = no answer. 404 means the
    // version isn't signed; 5xx means the registry can't say; we
    // fall through either way.
    if !response.status().is_success() {
        return Ok(None);
    }
    let body: serde_json::Value = match response.json().await {
        Ok(body) => body,
        Err(error) => {
            tracing::debug!(target: "pacquet_resolving_npm_resolver::attestation", ?error, %url, "attestation body parse failed; falling back");
            return Ok(None);
        }
    };
    Ok(extract_published_at(&body))
}

/// Pull the earliest `integratedTime` across every attestation
/// bundle in the response and convert it to an ISO timestamp.
/// Earliest is the conservative choice: if two attestations
/// disagree (e.g. publish v0.1 vs SLSA provenance v1) the older
/// Rekor entry is what tells us when the artifact existed in a
/// transparency log, which is the floor on publish time.
fn extract_published_at(body: &serde_json::Value) -> Option<String> {
    let attestations = body.get("attestations")?.as_array()?;
    let mut earliest: Option<i64> = None;
    for attestation in attestations {
        let Some(seconds) = read_earliest_integrated_time(attestation) else { continue };
        earliest = Some(earliest.map_or(seconds, |current| current.min(seconds)));
    }
    let seconds = earliest?;
    // `DateTime::from_timestamp` accepts an i64 of seconds-since-epoch.
    // Any seconds value within the chrono-supported range survives the
    // conversion; the registry isn't going to send pre-1970 timestamps.
    let dt = DateTime::from_timestamp(seconds, 0)?;
    Some(dt.to_rfc3339_opts(chrono::SecondsFormat::Millis, true))
}

fn read_earliest_integrated_time(attestation: &serde_json::Value) -> Option<i64> {
    let bundle = attestation.get("bundle")?;
    let verification_material = bundle.get("verificationMaterial")?;
    let tlog_entries = verification_material.get("tlogEntries")?.as_array()?;
    let mut earliest: Option<i64> = None;
    for entry in tlog_entries {
        // npm serializes integratedTime as a string ("1778583836")
        // to avoid JSON precision loss; accept either string or
        // number defensively.
        let seconds = parse_integrated_time_seconds(entry.get("integratedTime")?)?;
        earliest = Some(earliest.map_or(seconds, |current| current.min(seconds)));
    }
    earliest
}

fn parse_integrated_time_seconds(value: &serde_json::Value) -> Option<i64> {
    if let Some(text) = value.as_str() {
        return text.parse::<i64>().ok().filter(|&seconds| seconds > 0);
    }
    if let Some(seconds) = value.as_i64() {
        return Some(seconds).filter(|&s| s > 0);
    }
    None
}

#[cfg(test)]
mod tests;
